wasm-crypto/js-bindings/crypto_test.js

```diff
@@ -113,4 +122,137 @@ export async function run_test() {
+
+    const vrf_public_key = "tvrfpk1qpk0t6np4gyl084fv328h6ahjvwcsaktrzfrs0xeqtrzpp0l7p28knrnn57";
+
+    try {
+      const invalid_margin_ratio_per_thousand = 2000;
+      encode_stake_pool_data("40000", address, vrf_public_key, address, invalid_margin_ratio_per_thousand, "0", Network.Testnet);
+      throw new Error("Invalid margin_ratio_per_thousand worked somehow!");
+    } catch (e) {
+      if (!e.includes("Invalid per thousends 2000 valid range is [0, 1000]")) {
+        throw e;
+      }
+      console.log("Tested invalid margin_ratio_per_thousand successfully");
+    }
+
+    const pool_data = encode_stake_pool_data("40000", address, vrf_public_key, address, 100, "0", Network.Testnet);
+
+    const pool_id = "tpool1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqza035u";
+    try {
+      const invalid_pool_data = "invalid pool data";
+      encode_output_create_stake_pool(pool_id, invalid_pool_data, Network.Testnet);
+      throw new Error("Invalid pool data worked somehow!");
+    } catch (e) {
+      if (!e.includes("Invalid stake pool data encoding")) {
+        throw e;
+      }
+      console.log("Tested invalid pool data successfully");
+    }
+    const stake_pool_output = encode_output_create_stake_pool(pool_id, pool_data, Network.Testnet);
+    const outputs = [...output, ...stake_pool_output];
+
+    try {
+      const invalid_inputs = "invalid inputs";
+      encode_transaction(invalid_inputs, outputs, BigInt(0));
+      throw new Error("Invalid inputs worked somehow!");
+    } catch (e) {
+      if (!e.includes("Invalid Transaction input encoding")) {
+        throw e;
+      }
+      console.log("Tested invalid inputs successfully");
+    }
+
+    try {
+      const invalid_outputs = "invalid outputs";
+      encode_transaction(inputs, invalid_outputs, BigInt(0));
+      throw new Error("Invalid outputs worked somehow!");
+    } catch (e) {
+      if (!e.includes("Invalid Transaction output encoding")) {
+        throw e;
+      }
+      console.log("Tested invalid outputs successfully");
+    }
+
+    const tx = encode_transaction(inputs, outputs, BigInt(0));
+
+    const witness = encode_witness_no_signature();
+
+    try {
+      const invalid_witnesses = "invalid witnesses";
+      encode_signed_transaction(tx, invalid_witnesses);
+      throw new Error("Invalid witnesses worked somehow!");
+    } catch (e) {
+      if (!e.includes("Invalid Transaction witness encoding")) {
+        throw e;
+      }
+      console.log("Tested invalid witnesses successfully");
+    }
+
+    try {
+      encode_signed_transaction(tx, witness);
+      throw new Error("Invalid number of witnesses worked somehow!");
+    } catch (e) {
+      if (!e.includes("The number of signatures does not match the number of inputs")) {
+        throw e;
+      }
+      console.log("Tested invalid number of witnesses successfully");
+    }
+
+    try {
+      const invalid_tx = "invalid tx";
+      encode_signed_transaction(invalid_tx, witness);
```

Comment:
Please add chosen-text tests for serialization to ensure the encoding format never changes.

Reply:
This applies to all `encode_` functions